docs/RFCS/00000000_max_safe_timestamp.md

```diff
@@ -0,0 +1,241 @@
+  we'll call the `max_write_timestamp`.
+* The leaseholder will also track the `max_proposed_write_timestamp`.
+* The leaseholder promises not to propose a write with timestamp less than
+  `max_proposed_write_timestamp - max_write_age`, where `max_write_age` is a
+  configurable setting that will limit how old writes can be. The default
+  value for `max_write_age` will be in the tens of seconds.
+  * Making this safe will require taking these timestamps into account when
+    proposing or re-proposing commands.
+  * Making this safe may also require the leaseholder to propose empty commands
+    to make sure that writes which have been proposed but have not yet committed
+    will fail the `LeaseAppliedIndex` / `MaxLeaseIndex` check if they do
+    eventually commit. This would be to prevent a write from lingering for too
+    long and being older than `max_write_timestamp - max_write_age` by the time
+    it commits.
+* Followers can serve reads at times less than their `max_write_timestamp -
+  max_write_age`.
+  * This may mean the some followers can serve reads that others
+    can't, but all such reads will be correct as long as the leaseholder doesn't
+    break its promise about not proposing old writes.
+  * Intents encountered by the followers can be resolved as necessary. This will
+    add some additional latency to such reads, but maintains correctness. Any
+    intents left by transactions that are still pending should have their
+    transactions aborted at this point.
+* If a follower receives a stale read request that it can't serve, it has to
+  forward it to the leaseholder via a NotLeaseHolderError.
+* To make all of this usable, the `DistSender` will have to be instrumented to
+  decide which replica is closest, decide whether a follower may be able to
+  handle the request, and forward such requests to the leaseholder if the
+  follower can't.
+* The cluster's `maxOffset` doesn't need to be considered during steady-state
+  operation under a single range lease, because all timestamps for writes are
+  getting verified by the same leaseholder. `maxOffset` only needs to come in
+  play when the lease for a range changes hands, as described below.
+
+There are still a couple big questions left.
+
+As long as new writes are happening on the range, all the replicas will
+naturally have their `max_write_timestamp` kept up-to-date by the timestamps on
+the newly proposed writes. However, if no write has been proposed for a while,
+the followers' `max_write_timestamp` will fall far behind the actual current
+time, limiting the usefulness of follower reads in inactive ranges. This
+suggests that we need periodic "heartbeats" of some sort to keep followers
+up-to-date with respect to which timestamps are safe to read. There are two
+options here:
+
+1. Have the leaseholder periodically propose an empty command to raft with the
+   current timestamp if no other writes have come through lately.
```

Comment:
There could be another option to have the leaseholder periodically propose an empty command to raft with a future timestamp and disallow writes to that range until the future timestamp. This is like giving out a lease to the followers until a certain timestamp to allow reads at the current timestamp from followers at the expense of slowing down writes considerably